src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -249,70 +206,135 @@ protected virtual string GenerateUniqueId()
+
+                // Bound attributes have associated descriptors.
+                if (associatedDescriptors.Any())
+                {
+                    if (!renderedBoundAttributeNames.Add(attribute.Key))
                     {
+                        // First attribute value wins if there are duplicates.
                         continue;
                     }
 
-                    // We need to inform the context of the attribute value.
-                    _writer
-                        .WriteStartInstanceMethodInvocation(
-                            ExecutionContextVariableName,
-                            _tagHelperContext.ExecutionContextAddTagHelperAttributeMethodName)
-                        .WriteStringLiteral(attributeName)
-                        .WriteParameterSeparator()
-                        .Write(valueAccessor)
-                        .WriteEndMethodInvocation();
+                    // We need to capture the tag helper's property value accessor so we can retrieve it later
+                    // if there are more tag helpers that need the value.
+                    string valueAccessor = null;
+
+                    foreach (var associatedDescriptor in associatedDescriptors)
+                    {
+                        var associatedAttributeDescriptor = associatedDescriptor.Attributes.First(
+                            attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key));
+                        var tagHelperVariableName = GetVariableName(associatedDescriptor);
+
+                        valueAccessor = RenderBoundAttribute(
+                            attribute.Key,
+                            attribute.Value,
+                            tagHelperVariableName,
+                            valueAccessor,
+                            associatedAttributeDescriptor);
+                    }
                 }
                 else
                 {
-                    // The attribute value has already been recorded, lets retrieve it from the stored value
-                    // accessors.
-                    _writer
-                        .WriteStartAssignment(valueAccessor)
-                        .Write(htmlAttributeValues[attributeName])
-                        .WriteLine(";");
+                    RenderUnboundAttribute(attribute.Key, attribute.Value);
+                }
+            }
+        }
+
+        private string RenderBoundAttribute(
+            string attributeName,
+            Chunk attributeValueChunk,
+            string tagHelperVariableName,
+            string previousValueAccessor,
+            TagHelperAttributeDescriptor attributeDescriptor)
+        {
+            if (attributeValueChunk == null)
+            {
+                // Minimized attributes are not valid for bound attributes. TagHelperBlockRewriter has already
+                // logged an error if it was a bound attribute; so we can skip.
+                return null;
```

Comment:
Side note: If an attribute name is duplicated and the first use is minimized, will never render anything for it. Shouldn't matter since this is an error scenario but worth keeping in mind if we ever support minimized bound attributes. About the worst aspect is that errors in the non-minimized attribute value won't show up until the minimized one is removed.

E.g. with
``` html
  <tagHelper bound-attribute bound-attribute="value" />
```
Nothing will be generated for `bound-attribute`. Well, perhaps dictionary initialization (if the binding is to an indexer with a name such as `bound-`) but that's terribly minor.